### PR TITLE
[Feature][Executor] Support line timeout in async batch exec_line

### PR DESCRIPTION
**Pull request by Jasmin3q**
# Description

Support line timeout in exec_line to ensure run_info is also persisted in batch run.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py [hunk 1]
```diff
@@ -76,7 +99,30 @@ async def execute(
         # This is because it will always call `executor.shutdown()` when exiting the `with` block.
         # Then the event loop will wait for all tasks to be completed before raising the cancellation error.
         # See reference: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor
-        outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
+        try:
+            if line_timeout_sec is not None:
+                outputs = await self._execute_with_timeout(executor, nodes, inputs, context, line_timeout_sec)
+            else:
+                outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
```
Comment (guming-learning) [inline src/promptflow/promptflow/executor/_async_nodes_scheduler.py:106]: Please rename this function to `self._execute`

You see that this is weird that:
```
if line_timeout_sec is not None:
    self._execute_with_timeout
else:
    self._execute_with_thread_pool
```
But this can be another PR

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py [hunk 2]
```diff
@@ -12,17 +12,22 @@
...
 from promptflow._core.tools_manager import ToolsManager
 from promptflow._utils.logger_utils import flow_logger
 from promptflow._utils.utils import extract_user_frame_summaries, set_context
 from promptflow.contracts.flow import Node
 from promptflow.executor._dag_manager import DAGManager
-from promptflow.executor._errors import NoNodeExecutedError
+from promptflow.executor._errors import LineExecutionTimeoutError, NoNodeExecutedError
 
 PF_ASYNC_NODE_SCHEDULER_EXECUTE_TASK_NAME = "_pf_async_nodes_scheduler.execute"
+PF_ASYNC_NODE_SCHEDULER_EXECUTE_WITH_THREAD_POOL = "_pf_async_nodes_scheduler._execute_with_thread_pool"
```
Comment (guming-learning) [inline src/promptflow/promptflow/executor/_async_nodes_scheduler.py:26]: Will the name of scheduler task change after your change? Or your change will introduce another task?

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py [hunk 3]
```diff
@@ -76,7 +99,30 @@ async def execute(
...
+        except Exception as e:
+            err_msg = "Flow execution has failed."
+            if isinstance(e, LineExecutionTimeoutError):
+                err_msg = f"Line execution timeout after {line_timeout_sec} seconds."
+                context.cancel_node_runs(err_msg)
+            running_tasks = [task for task in asyncio.all_tasks(loop) if not task.done()]
+            node_names = ",".join(
+                task.get_name()
+                for task in running_tasks
+                if task.get_name() not in PF_ASYNC_NODE_SCHEDULER_EXECUTE_TASKS
```
Comment (guming-learning) [inline src/promptflow/promptflow/executor/_async_nodes_scheduler.py:116]: It seems we need to maintain a list of running task by ourself in this class, it would be unmaintainable to exclude so many task names. Let's do it in another PR.
I.e., keep a dictionary of {node_name: task}, add when create task, and del key when the task is done. Then we can avoid scanning eventloop

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py (hunk 2 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/executor/_async_nodes_scheduler.py:26]: The asyncio.wait_for will introduce another task. If we don't set the name of this task, it will raise warning during the process.

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py (hunk 1 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/executor/_async_nodes_scheduler.py:106]: Changed to self._execute.

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py [hunk 4]
```diff
@@ -76,7 +95,28 @@ async def execute(
...
+            else:
+                outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
+        except Exception as e:
+            err_msg = "Flow execution has failed."
+            if isinstance(e, LineExecutionTimeoutError):
+                err_msg = f"Line execution timeout after {line_timeout_sec} seconds."
+                context.cancel_node_runs(err_msg)
+            node_names = ",".join(task.get_name() for task in self._running_tasks.values())
+            flow_logger.error(f"{err_msg} Cancelling all running nodes: {node_names}.")
+            current_running_task_names = list(self._running_tasks.keys())
```
Comment (guming-learning) [inline src/promptflow/promptflow/executor/_async_nodes_scheduler.py:110]: Seems no need to create a list of keys? It should be iterable.

Comment (guming-learning) [reply src/promptflow/promptflow/executor/_async_nodes_scheduler.py:110]: Offline discussed, it would cause dict size change during iteration, let's keep the current way.
Maybe use a more concise code, sth like:
```
for name in for name in current_running_task_names:
```
But it's optional

**Review by guming-learning — APPROVED** [review]

File: src/promptflow/tests/executor/e2etests/test_batch_timeout.py [hunk 5]
```diff
@@ -170,3 +171,52 @@ def test_batch_timeout(self, flow_folder):
...
         # We will assert the correct result after refining the implementation of batch timeout.
         assert len(mem_run_storage._node_runs) == 5, "Node runs are persisted in memory storage."
+
+    @pytest.mark.parametrize(
+        "flow_folder",
+        [
+            ASYNC_SAMPLE_FLOW,
+        ],
+    )
+    def test_async_batch_with_line_timeout(self, flow_folder, dev_connections):
```
Comment (thy09) [inline src/promptflow/tests/executor/e2etests/test_batch_timeout.py:181]: We may also test the case that includes both async/sync tasks in the async scheduler?

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py [hunk 6]
```diff
@@ -76,7 +95,28 @@ async def execute(
         # This is because it will always call `executor.shutdown()` when exiting the `with` block.
         # Then the event loop will wait for all tasks to be completed before raising the cancellation error.
         # See reference: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor
-        outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
+        try:
+            if line_timeout_sec is not None:
+                outputs = await self._execute(executor, nodes, inputs, context, line_timeout_sec)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_async_nodes_scheduler.py:100]: Maybe renamed to `_execute_with_timeout` and `_execute`?

**Review by guming-learning — DISMISSED** [review]

File: src/promptflow/promptflow/executor/_async_nodes_scheduler.py (hunk 4 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/executor/_async_nodes_scheduler.py]: In the same way, we change the `for task in running_tasks.values():` to `for task in list(running_tasks.values()):` to avoid the same problem.

File: src/promptflow/tests/executor/e2etests/test_batch_timeout.py (hunk 5 above)
Comment (Jasmin3q) [reply src/promptflow/tests/executor/e2etests/test_batch_timeout.py]: After adding the test case "async_tools_with_sync_tools", we can come to these conclusions:
1. The async nodes and sync nodes will both be cancelled successfully.


2. The handing of CancelledError in node async_passthrough will cause a new problem:
27eb2a1a596f9256286d13b33545234
This is weird and need further investivation.
Currently we added "if" conditon to stdout to mitigate it. The behavior is like, after timeout, some of the CancelledError handing process output will loss and not demonstrated in stdout.
